### PR TITLE
Make MultistreamMuxer and Client APIs generic

### PR DESCRIPTION
**Pull request by sukunrt**
This change allows users to use their custom types instead of translating to and from string type

Related to: https://github.com/libp2p/go-libp2p/issues/1886

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -152,7 +152,7 @@ func SelectWithSimopenOrFail(protos []string, rwc io.ReadWriteCloser) (proto str
 
 func clientOpen(protos []string, rwc io.ReadWriteCloser) (string, error) {
 	// check to see if we selected the pipelined protocol
-	tok, err := ReadNextToken(rwc)
+	tok, err := ReadNextToken[string](rwc)
```
Comment (marten-seemann) [inline client.go:155]: Should this function also be generic?

File: client.go [hunk 2]
```diff
@@ -339,6 +339,6 @@ func readProto(proto string, r io.Reader) error {
 	case "na":
 		return ErrNotSupported
 	default:
-		return errors.New("unrecognized response: " + tok)
+		return errors.New("unrecognized response: " + string(tok))
```
Comment (marten-seemann) [inline client.go:342]: Nit: `fmt.Errorf` would probably be nicer.

**Review by marten-seemann — COMMENTED** [review]
This looks very clean. I like it!

I added a few comments.

File: client.go (hunk 1 above)
Comment (sukunrt) [reply client.go:155]: yes, it'd need to be. Missed this.

File: client.go (hunk 2 above)
Comment (sukunrt) [reply client.go]: fixed.

**Comment by sukunrt** [comment]
Addressed review comments.

**Review by marten-seemann — COMMENTED** [review]
LGTM. Just for completeness, should NewMultistream also be made generic?

**Comment by sukunrt** [comment]
Agreed. I think this now covers the complete API.

**Review by marten-seemann — APPROVED** [review]

**Comment by marten-seemann** [comment]
Nice work, @sukunrt!

**Pull request merged**